File: python/pondsocket/src/pondsocket/parser.py
from __future__ import annotations

from urllib.parse import parse_qs, unquote

from .errors import not_found
from .types import Route


def parse(route: str, current_path: str) -> Route:
    query: dict[str, list[str]] = {}
    params: dict[str, str] = {}
    wildcard: str | None = None

    if current_path == "":
        current_path = "/"

    if "?" in current_path:
        path, _, query_string = current_path.partition("?")
        query = parse_qs(query_string, keep_blank_values=True)
    else:
        path = current_path

    route_segments = _split_path(route)
    path_segments = _split_path(path)

    wildcard_index = -1
    for i, seg in enumerate(route_segments):
        if seg == "*":
            wildcard_index = i
            break

    matched = False
    if wildcard_index >= 0:
        if wildcard_index < len(path_segments):
            remaining = "/".join(path_segments[wildcard_index:])
            wildcard = unquote(remaining)
        else:
            wildcard = ""
        route_match = route_segments[:wildcard_index]
        path_match = path_segments[: min(wildcard_index, len(path_segments))]
        matched = _match_segments(route_match, path_match, params)
    else:
        if len(route_segments) == len(path_segments):
            matched = _match_segments(route_segments, path_segments, params)

    if not matched:
        raise not_found(route, f"route {route} does not match path {current_path}")

    return Route(params=params, query=query, wildcard=wildcard)


def _match_segments(
    route_segments: list[str],
    path_segments: list[str],
    params: dict[str, str],
) -> bool:
    if len(route_segments) > len(path_segments):
        return False
    for i, route_seg in enumerate(route_segments):
        if i >= len(path_segments):
            return False
        path_seg = path_segments[i]
        if route_seg.startswith(":"):
            param_name = route_seg[1:]
            params[param_name] = unquote(path_seg)
            continue
        if ":" in route_seg:
            prefix, _, param_name = route_seg.partition(":")
            if not path_seg.startswith(prefix):
                return False
            value = path_seg[len(prefix) :]
            params[param_name] = unquote(value)
            continue
        if route_seg != path_seg:
            return False
    return True


def _split_path(path: str) -> list[str]:
    path = path.strip("/")
    while "//" in path:
        path = path.replace("//", "/")
    if path == "":
        return []
    return path.split("/")

File: rust/pondsocket/src/parser.rs
use std::collections::HashMap;

use crate::errors::{Result, not_found};
use crate::types::Route;

pub fn parse(route: &str, current_path: &str) -> Result<Route> {
    let mut params = HashMap::new();
    let mut query = HashMap::new();
    let mut wildcard = None;
    let current_path = if current_path.is_empty() {
        "/"
    } else {
        current_path
    };
    let (path, query_string) = current_path
        .split_once('?')
        .map_or((current_path, ""), |(p, q)| (p, q));

    if !query_string.is_empty() {
        for part in query_string.split('&') {
            let (key, value) = part.split_once('=').unwrap_or((part, ""));
            query
                .entry(percent_decode(key))
                .or_insert_with(Vec::new)
                .push(percent_decode(value));
        }
    }

    let route_segments = split_path(route);
    let path_segments = split_path(path);
    let wildcard_index = route_segments.iter().position(|s| s == "*");

    let matched = if let Some(index) = wildcard_index {
        wildcard = Some(path_segments.get(index..).unwrap_or(&[]).join("/"));
        match_segments(
            &route_segments[..index],
            &path_segments[..index.min(path_segments.len())],
            &mut params,
        )
    } else {
        route_segments.len() == path_segments.len()
            && match_segments(&route_segments, &path_segments, &mut params)
    };

    if !matched {
        return Err(not_found(
            route,
            format!("route {route} does not match path {current_path}"),
        ));
    }

    Ok(Route {
        params,
        query,
        wildcard,
    })
}

fn match_segments(
    route_segments: &[String],
    path_segments: &[String],
    params: &mut HashMap<String, String>,
) -> bool {
    if route_segments.len() > path_segments.len() {
        return false;
    }
    for (route_seg, path_seg) in route_segments.iter().zip(path_segments.iter()) {
        if let Some(param_name) = route_seg.strip_prefix(':') {
            params.insert(param_name.to_owned(), percent_decode(path_seg));
            continue;
        }
        if let Some((prefix, param_name)) = route_seg.split_once(':') {
            if !path_seg.starts_with(prefix) {
                return false;
            }
            params.insert(
                param_name.to_owned(),
                percent_decode(&path_seg[prefix.len()..]),
            );
            continue;
        }
        if route_seg != path_seg {
            return false;
        }
    }
    true
}

fn split_path(path: &str) -> Vec<String> {
    let mut path = path.trim_matches('/').replace("//", "/");
    while path.contains("//") {
        path = path.replace("//", "/");
    }
    if path.is_empty() {
        Vec::new()
    } else {
        path.split('/').map(ToOwned::to_owned).collect()
    }
}

fn percent_decode(value: &str) -> String {
    let mut out = String::with_capacity(value.len());
    let bytes = value.as_bytes();
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i] == b'%'
            && i + 2 < bytes.len()
            && let Ok(hex) = u8::from_str_radix(&value[i + 1..i + 3], 16)
        {
            out.push(hex as char);
            i += 3;
            continue;
        }
        out.push(bytes[i] as char);
        i += 1;
    }
    out
}

#[cfg(test)]
mod tests {
    use super::parse;

    #[test]
    fn matches_params_prefix_and_query() {
        let route = parse("/chat/room-:id/*", "/chat/room-42/a/b?token=x&token=y").unwrap();
        assert_eq!(route.param("id"), Some("42"));
        assert_eq!(route.wildcard.as_deref(), Some("a/b"));
        assert_eq!(
            route.query_param("token"),
            &["x".to_string(), "y".to_string()]
        );
    }
}
